proxy/http2/Http2ConnectionState.cc

```diff
@@ -1067,16 +1067,33 @@ Http2ConnectionState::find_stream(Http2StreamId id) const
 void
 Http2ConnectionState::restart_streams()
 {
-  Http2Stream *s = stream_list.head;
-
-  while (s) {
-    Http2Stream *next = static_cast<Http2Stream *>(s->link.next);
+  static int x = 0;
+
+  Http2Stream *s   = stream_list.head;
+  Http2Stream *end = s;
+  if (s) {
+    // Change the start point randomly
+    for (int i = x % total_client_streams_count; i; --i) {
+      end = static_cast<Http2Stream *>(end->link.next ? end->link.next : stream_list.head);
+    }
+    s = static_cast<Http2Stream *>(end->link.next ? end->link.next : stream_list.head);
+
+    // Call send_response_body() for each streams
+    while (s != end) {
+      Http2Stream *next = static_cast<Http2Stream *>(s->link.next ? s->link.next : stream_list.head);
+      if (!s->is_closed() && s->get_state() == Http2StreamState::HTTP2_STREAM_STATE_HALF_CLOSED_REMOTE &&
+          std::min(this->client_rwnd, s->client_rwnd) > 0) {
+        s->send_response_body();
+      }
+      ink_assert(s != next);
+      s = next;
+    }
     if (!s->is_closed() && s->get_state() == Http2StreamState::HTTP2_STREAM_STATE_HALF_CLOSED_REMOTE &&
         std::min(this->client_rwnd, s->client_rwnd) > 0) {
       s->send_response_body();
     }
-    ink_assert(s != next);
-    s = next;
+
+    x = (x + 1) & 0xFFFF;
   }
```

Comment:
Oh, I guess I see now.  X is a static/global and will get incremented in subsequent calls.  May want to add a comment someplace to make this obvious

Reply:
Right. I'll add a comment.

Reply:
We should do a do while to combine the code in the loop and the conditional below.